homeassistant/components/mobile_app/entity.py

```diff
@@ -0,0 +1,135 @@
+        platform_name = "MobileAppBinarySensor"
+
+    platform = globals()[platform_name]
+
+    entities = list()
+
+    if discovery_info is None:
+        for sensor_config in sensor_configs.values():
+            device = registrations[sensor_config[CONF_WEBHOOK_ID]]
+            entities.append(platform(sensor_config, device))
+    else:
+        device = registrations[discovery_info[CONF_WEBHOOK_ID]]
+        sensor_config = sensor_configs[discovery_info[ATTR_SENSOR_UNIQUE_ID]]
+        entities.append(platform(sensor_config, device))
+
+    async_add_entities(entities, True)
+
+
+class MobileAppEntity(Entity):
+    """Representation of an mobile app entity."""
+
+    def __init__(self, config, device):
+        """Initialize the sensor."""
+        self._config = config
+        self._device = device
+        self._device_id = device[CONF_WEBHOOK_ID]
+        self._sensor_id = "{}_{}".format(device[CONF_WEBHOOK_ID],
+                                         config[ATTR_SENSOR_UNIQUE_ID])
+        self._entity_type = config[ATTR_SENSOR_TYPE]
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+        async_dispatcher_connect(self.hass, SIGNAL_SENSOR_UPDATE,
+                                 self._handle_update)
+
+    async def async_will_remove_from_hass(self):
+        """Disconnect dispatcher listener when removed."""
+        async_dispatcher_connect(self.hass, SIGNAL_SENSOR_UPDATE,
+                                 self._handle_update)
+
+    @property
+    def should_poll(self) -> bool:
+        """Declare that this entity pushes its state to HA."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the mobile app sensor."""
+        return self._config[ATTR_SENSOR_NAME]
+
+    @property
+    def device_class(self):
+        """Return the device class."""
+        return self._config.get(ATTR_SENSOR_DEVICE_CLASS)
+
+    @property
+    def device_state_attributes(self):
+        """Return the device state attributes."""
+        return self._config[ATTR_SENSOR_ATTRIBUTES]
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        return self._config[ATTR_SENSOR_ICON]
+
+    @property
+    def unique_id(self):
+        """Return the unique ID of this sensor."""
+        return self._sensor_id
+
+    @property
+    def is_on(self):
+        """Return the state of the binary sensor."""
+        return self._config[ATTR_SENSOR_STATE]
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        if self._entity_type == ATTR_SENSOR_TYPE_BINARY_SENSOR:
+            return STATE_ON if self.is_on else STATE_OFF
+
+        return self._config[ATTR_SENSOR_STATE]
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement this sensor expresses itself in."""
+        if self._entity_type == ATTR_SENSOR_TYPE_BINARY_SENSOR:
```

Comment:
I would suggest separating the non binary sensor from the binary sensor with two different subclasses, to avoid the checks in the properties.

Reply:
@MartinHjelmare I understand why you'd suggest separating them but at the same time there will be so much unnecessary logic duplication IMHO. I think it's better to have these conditionals instead of duplicate logic. Thoughts?

Reply:
The only methods we need to implement in the child classesaare the ones where the methods will differ between the child classes.

Reply:
Will split.